### PR TITLE
Implement array_min and array_max scalar functions.

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java:85]: Only a few of the value instances implement `Comparable`, so this might fail at runtime for many of the supported type.

`DataType` has a `compare` function that could be used instead to support more types. Might also make sense to restrict the function registration to only match on types which we actually want to support.

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java]: Good point, thanks.

>Restrict function registration:

Iterating over PRIMITIVE_TYPES now - similar to aggregation min/max - I believe supported type set must be equivalent.

>might fail at runtime

Updated _happy path test_ to check support of all PRIMITIVE_TYPES.

>`DataType` has a `compare` function

Done. However, not sure about usage of PRIMITIVE_TYPES now - I noticed non-primitive types with compare (GeoPoint, Regclass) - on the one hand, why not support them as they have compare, on the other hand, is it fine that aggregation `max` and `array_max` would have different argument type set?

File: server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java:83]: Could you move this to the constructor? `createType()` causes allocations, so it would be good to having them on each `evaluate` call (which is per row)

File: server/src/main/java/io/crate/expression/scalar/ArrayMinFunction.java
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/scalar/ArrayMinFunction.java:83]: Same here

File: server/src/test/java/io/crate/expression/scalar/ArrayMinFunctionTest.java
Comment (mfussenegger) [inline server/src/test/java/io/crate/expression/scalar/ArrayMinFunctionTest.java:21]: I wonder if we should instead extend `assertEvaluate` to also support parameter placeholders, so that you could do something like `assertEvaluate("array_min(?)", expected, Literal.of([1, 2, 3]))`

It already supports column references, but it only works for the columns that are defined in the base class (see `prepareFunctions` in `AbstractScalarFunctionsTest`)

File: docs/appendices/release-notes/unreleased.rst
Comment (norosa) [inline docs/appendices/release-notes/unreleased.rst:69]: these two items could probably be collapsed into a single item

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2457]:
```suggestion
The ``array_min`` function returns the smallest element in ``array``. If
``array`` is ``NULL`` or an empty array, the function returns``NULL``.
```
not sure what this line means:

> Return value is always of the type of input array.

I presume it means the same type as the smallest element? if it does mean that, then I think that is implied and doesn't need stating. we typically only explicitly mention it if a type cast _is_ performed

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2452]:
```suggestion
``array_min(array)``
--------------------
```

switched the argument to ``array`` for improved clarity and to match the subsequent text

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2461]:
```suggestion
    cr> SELECT array_min([3, 2, 1]) AS min;
```

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2472]:
```suggestion
``array_max(array)``
--------------------
```

as above

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2477]:
```suggestion
The ``array_max`` function returns the largest element in ``array``. If
``array`` is ``NULL`` or an empty array, the function returns``NULL``.
```

as above

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2481]:
```suggestion
    cr> SELECT array_max([1,2,3]) AS max;
```

**Review by norosa — CHANGES_REQUESTED** [review]
thanks for the PR! left some minor requested changes

File: server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/expression/scalar/ArrayMaxFunction.java:83]: Done

File: server/src/main/java/io/crate/expression/scalar/ArrayMinFunction.java
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/expression/scalar/ArrayMinFunction.java:83]: Done

File: docs/appendices/release-notes/unreleased.rst
Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/unreleased.rst:69]: Done

File: server/src/test/java/io/crate/expression/scalar/ArrayMinFunctionTest.java
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/expression/scalar/ArrayMinFunctionTest.java]: Done (exclusive implementation, Literals are either params or refs, if we want both then probably need another overload 
with 2 Lists -one for params, one for refs, not sure we have such test case).

Facing casting exceptions, something to do with module.register. Converted PR to draft for now

File: docs/appendices/release-notes/unreleased.rst
Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/unreleased.rst]: @norosa @mfussenegger should I mention here, that only primitive types are supported? We have some types with implemented `compare`, for example `GeoPointType` and users might have an impression that they can find min/max for those types.

File: server/src/test/java/io/crate/expression/scalar/ArrayMinFunctionTest.java
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/expression/scalar/ArrayMinFunctionTest.java]: Done

File: docs/appendices/release-notes/unreleased.rst
Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/unreleased.rst]: Added note about primitive types in scalar-functions.rst

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2456]:
```suggestion
The ``array_min`` function returns the smallest element in ``array``. If
``array`` is ``NULL`` or an empty array, the function returns ``NULL``. This
function supports arrays of any of the :ref:`primitive types
<sql_ddl_datatypes_primitives>`.
```

- corrected missing space before `NULL`
- added "This function" so the sentence is in the active voice (i.e., the do-er is explicit)
- added explicit link text for `sql_ddl_datatypes_primitives` (it's best to avoid relying on the heading text of the link target being or remaining suitable)
- removed the double space (`  `) after the period (`.`) because spaces are collapsed by RST so double spaces only introduce diff noise
- rewrapped to 79 chars

File: docs/general/builtins/scalar-functions.rst
Comment (norosa) [inline docs/general/builtins/scalar-functions.rst:2475]:
```suggestion
The ``array_max`` function returns the largest element in ``array``. If
``array`` is ``NULL`` or an empty array, the function returns ``NULL``. This
function supports arrays of any of the :ref:`primitive types
<sql_ddl_datatypes_primitives>`.
```

as above

**Review by norosa — CHANGES_REQUESTED** [review]

File: docs/general/builtins/scalar-functions.rst
Comment (BaurzhanSakhariev) [reply docs/general/builtins/scalar-functions.rst:2456]: Thanks for corrections!

File: docs/appendices/release-notes/unreleased.rst
Comment (norosa) [inline docs/appendices/release-notes/unreleased.rst:66]:
```suggestion
- Added :ref:`array_min <scalar-array-min>` and :ref:`array_max
  <scalar-array-max>` scalar functions returning the minimal and maximal
  element in array respectively.
```

**Review by norosa — APPROVED** [review]
spotted a typo! otherwise lgtm!

I am adding an approval so my "request changes" is removed from the pr. however, please do not merge until you have an approval from an engineer (my approvals do not apply to your code changes)

File: server/src/main/java/io/crate/expression/symbol/ParameterBinder.java
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/symbol/ParameterBinder.java:29]: Maybe we could move this to the `testFixtures` as it's currently only used in the tests.

File: server/src/test/java/io/crate/expression/scalar/ArrayMaxFunctionTest.java
Comment (mfussenegger) [inline server/src/test/java/io/crate/expression/scalar/ArrayMaxFunctionTest.java:30]:
```suggestion
            String expression = String.format(Locale.ENGLISH, "array_max(?::%s[])", type.getName());
```

We randomize the locales in the tests - I don't think it is an issue in this case, but it's a good habit to declare the locale explicitly.
Same in the ArrayMinFunctionTest

(I thought our forbidden-apis check would even complain about this)

File: server/src/testFixtures/java/io/crate/testing/TestingHelpers.java
Comment (mfussenegger) [inline server/src/testFixtures/java/io/crate/testing/TestingHelpers.java:452]: I think it might make sense to do the `.add(null)` on a random basis as well, to ensure arrays without any null values are also tested.

And I think it shouldn't be added as an additional element. That the list that is returned can exceeds `maxLength` could be surprising.

**Review by mfussenegger — COMMENTED** [review]
Left some more minor comments. Should be good to go after they're addressed. :+1:

File: server/src/testFixtures/java/io/crate/testing/TestingHelpers.java
Comment (BaurzhanSakhariev) [reply server/src/testFixtures/java/io/crate/testing/TestingHelpers.java:452]: loop is to `length - 1` and if  `maxLength<2` Exception with "..One item is reserved for NULL." is thrown  - null case can be random indeed, will add that, thanks

File: server/src/testFixtures/java/io/crate/testing/TestingHelpers.java
Comment (mfussenegger) [inline server/src/testFixtures/java/io/crate/testing/TestingHelpers.java:454]: What do you think of implementing it like this:


```suggestion
        var generator = DataTypeTesting.getDataGenerator(dataType);
        for (int i = 0; i < length; i++) {
            // 1/length chance
            if (RandomizedTest.randomInt(length) == 0) {
              values.add(null);
            } else {
              values.add(dataType.sanitizeValue(generator.get()));
            }
        }
```

This should still result in approximately 1 null value per array, but randomizes the position of it. So that it is not always at the start or end, but can be anywhere.

**Review by mfussenegger — APPROVED** [review]
Left another suggestion regarding the array randomization code, but it's also good as it is. Leave it up to you.

Please don't forget to squash before merging.

File: server/src/testFixtures/java/io/crate/testing/TestingHelpers.java
Comment (BaurzhanSakhariev) [reply server/src/testFixtures/java/io/crate/testing/TestingHelpers.java:454]: Oh, sorry, I saw that comment after adding ready-to-merge label. I can do this in array_sum PR.

**Pull request merged**
